Breaking change was introduced during PATCH change

Hi!

Afer upgrade  from `3.0.8` to `3.0.9` and above my current karma setup stop working with this meaningless message (in this case):

> You need to include some adapter that implements \_\_karma__.start method!

If we run `karma` without `--single-run` we will get a bit more information.

> Unhandled rejection Error: COMPILATION ERROR
>     at maybeWrapAsError (/projects/test/node_modules/bluebird/js/release/util.js:61:12)
>     at /projects/test/node_modules/bluebird/js/release/nodeback.js:38:50
>     at nextPreprocessor (/projects/test/node_modules/karma/lib/preprocessor.js:26:14)
>     at Object.callback (/projects/test/node_modules/karma-typescript/dist/karma/preprocessor.js:19:36)
>     at /projects/test/node_modules/karma-typescript/dist/compiler/compiler.js:73:20
>     at Array.forEach (<anonymous>:null:null)
>     at Compiler.onProgramCompiled (/projects/test/node_modules/karma-typescript/dist/compiler/compiler.js:66:24)
>     at Compiler.compileProject (/projects/test/node_modules/karma-typescript/dist/compiler/compiler.js:52:14)
>     at Compiler.<anonymous> (/projects/test/node_modules/karma-typescript/dist/compiler/compiler.js:28:23)
>     at invokeFunc (/projects/test/node_modules/karma-typescript/node_modules/lodash/lodash.js:10333:23)
>     at trailingEdge (/projects/test/node_modules/karma-typescript/node_modules/lodash/lodash.js:10382:18)
>     at Timeout.timerExpired [as _onTimeout] (/projects/test/node_modules/karma-typescript/node_modules/lodash/lodash.js:10370:18)
>     at ontimeout (timers.js:498:11)
>     at tryOnTimeout (timers.js:323:5)
>     at Timer.listOnTimeout (timers.js:290:5)

Diggin thru the change log I found #215 #216 which is a bit breaking change as you may see above.

To summarize this I kindly ask:
- Please provide a way to turn this off since by default TSC emits code (`noEmitOnError: false`)
- Don't introduce breaking changes in PATCH updates
- Fix message when karma runs with `--single-run` flag because "You need to include some adapter that implements \_\_karma__.start method!" is not sufficient

Thanks!

Used dependencies:

> "karma": "~2.0.4",
> "karma-babel-preprocessor": "~7.0.0",
> "karma-chrome-launcher": "^2.2.0",
> "karma-fixture": "0.2.6",
> "karma-jasmine": "~1.1.2",
> "karma-json-fixtures-preprocessor": "0.0.6",
> "karma-spec-reporter": "~0.0.28",
> "karma-typescript": "~3.0.12",

Related to #244 #249 #269

as soon as you use jasmine-matchers you won't have any success in running karma with karma-typescript, their typings throw errors (at least with ts 3.7.4):
> 19 02 2020 11:11:05.210:ERROR [compiler.karma-typescript]: node_modules/jasmine-expect/dist/toBeFalse.d.ts(5,13): error TS2717: Subsequent property declarations must have the same type.  Property 'toBeFalse' must be of type '() => boolean', but here has type 'ToBeFalse'.

how about a flag to ignore diagnostics? e.g.
```
karmaTypescriptConfig: {
  skipDiagnostics: true
}
```
and then within `compiler.ts`:
```
        [...]
        this.cachedProgram = this.program;
        if (!this.config.karmaTypescriptConfig.skipDiagnostics) {
            this.runDiagnostics(this.program, this.compilerHost);
        }
        this.program.emit();
        [...]
```

--- a/packages/karma-typescript-postcss-transform/src/transform.spec.ts
+++ b/packages/karma-typescript-postcss-transform/src/transform.spec.ts
@@ -89,10 +89,10 @@ test("transformer should set the source property to the processed value", (t) =>
     const context = createContext("::placeholder {}");
 
     transform(require("autoprefixer"))(context, () => {
-        t.isEqual(context.source, "::-webkit-input-placeholder {}\n" +
-                                  "::-moz-placeholder {}\n" +
+        t.isEqual(context.source, "::-moz-placeholder {}\n" +
                                   ":-ms-input-placeholder {}\n" +
-                                  "::-ms-input-placeholder {}\n::placeholder {}");
+                                  "::-ms-input-placeholder {}\n" + 
+                                  "::placeholder {}");
     });
 });
 
@@ -102,17 +102,15 @@ test("transformer should use custom options", (t) => {
     const context = createContext("::placeholder {}");
 
     transform(require("autoprefixer"), { map: { inline: true } })(context, () => {
-        t.isEqual(context.source, "::-webkit-input-placeholder {}\n" +
-                                  "::-moz-placeholder {}\n" +
+        t.isEqual(context.source, "::-moz-placeholder {}\n" +
                                   ":-ms-input-placeholder {}\n" +
                                   "::-ms-input-placeholder {}" +
                                   "\n::placeholder {}" +
-                                  "\n/*# sourceMappingURL=data:application/json;base64," +
-                                  "eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZpbGUuY3NzIl0sIm5hb" +
-                                  "WVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLDZCQUFlO0FBQWYsb0JBQW" +
-                                  "U7QUFBZix3QkFBZTtBQUFmLHlCQUFlO0FBQWYsZUFBZSIsImZpbGU" +
-                                  "iOiJmaWxlLmNzcyIsInNvdXJjZXNDb250ZW50IjpbIjo6cGxhY2Vo" +
-                                  "b2xkZXIge30iXX0= */");
+                                  "\n/*# sourceMappingURL=data:application/json;base64," + 
+                                  "eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZpbGUuY3NzIl0sIm5hbWV" +
+                                  "zIjpbXSwibWFwcGluZ3MiOiJBQUFBLG9CQUFlO0FBQWYsd0JBQWU7QU" +
+                                  "FBZix5QkFBZTtBQUFmLGVBQWUiLCJmaWxlIjoiZmlsZS5jc3MiLCJzb" + 
+                                  "3VyY2VzQ29udGVudCI6WyI6OnBsYWNlaG9sZGVyIHt9Il19 */");
     });
 });
 

--- a/packages/karma-typescript/README.md
+++ b/packages/karma-typescript/README.md
@@ -355,6 +355,10 @@ If the defaults aren't enough, the settings can be configured from `karma.conf.j
   The directory of the `tsconfig.json` file will be used as the base path for the Typescript compiler, and if `karmaTypescriptConfig.tsconfig` isn't set, the `basePath` property of the Karma config will be used as the
   compiler base path instead.
 
+* **karmaTypescriptConfig.stopOnFailure** - Stop on any compiler error.<br/>
+  By default karma will stop when any typescript compile errors are encountered.<br/>
+  Setting this to false will allow test tobe run when typescript compile errors are present.
+
 Example of a full `karmaTypescriptConfig` configuration:
 
 ```javascript
@@ -431,7 +435,8 @@ karmaTypescriptConfig: {
     transformPath: function(filepath) {
         return filepath.replace(/\.(ts|tsx)$/, ".js");
     },
-    tsconfig: "./tsconfig.json"
+    tsconfig: "./tsconfig.json",
+    stopOnFailure: false
 }
 ```
 

--- a/packages/karma-typescript/README.md
+++ b/packages/karma-typescript/README.md
@@ -357,7 +357,7 @@ If the defaults aren't enough, the settings can be configured from `karma.conf.j
 
 * **karmaTypescriptConfig.stopOnFailure** - Stop on any compiler error.<br/>
   By default karma will stop when any typescript compile errors are encountered.<br/>
-  Setting this to false will allow test tobe run when typescript compile errors are present.
+  Setting this to false will allow tests to be run when typescript compile errors are present.
 
 Example of a full `karmaTypescriptConfig` configuration:
 

--- a/packages/karma-typescript/src/api/configuration.ts
+++ b/packages/karma-typescript/src/api/configuration.ts
@@ -14,6 +14,7 @@ export interface KarmaTypescriptConfig {
     reports?: Reports;
     transformPath?: (filepath: string) => string;
     tsconfig?: string;
+    stopOnFailure?: boolean;
 }
 
 export interface BundlerOptions {

--- a/packages/karma-typescript/src/compiler/compiler.ts
+++ b/packages/karma-typescript/src/compiler/compiler.ts
@@ -106,7 +106,7 @@ export class Compiler {
 
             queued.callback({
                 ambientModuleNames,
-                hasError: this.errors.indexOf(queued.file.originalPath) !== -1,
+                hasError: this.config.stopOnFailure ? this.errors.indexOf(queued.file.originalPath) !== -1 : false,
                 isAmbientModule: ambientModuleNames && ambientModuleNames.length > 0,
                 isDeclarationFile: this.fileExtensionIs(sourceFile.fileName, ".d.ts"),
                 outputText: this.compiledFiles[queued.file.path],
@@ -150,7 +150,7 @@ export class Compiler {
 
     private outputDiagnostics(diagnostics: ReadonlyArray<ts.Diagnostic>, host?: ts.FormatDiagnosticsHost): void {
 
-        if (!diagnostics || diagnostics.length === 0 || !this.config.stopOnFailure) {
+        if (!diagnostics || diagnostics.length === 0) {
             return;
         }
 

--- a/packages/karma-typescript/src/compiler/compiler.ts
+++ b/packages/karma-typescript/src/compiler/compiler.ts
@@ -150,7 +150,7 @@ export class Compiler {
 
     private outputDiagnostics(diagnostics: ReadonlyArray<ts.Diagnostic>, host?: ts.FormatDiagnosticsHost): void {
 
-        if (!diagnostics || diagnostics.length === 0) {
+        if (!diagnostics || diagnostics.length === 0 || !this.config.stopOnFailure) {
             return;
         }
 

--- a/packages/karma-typescript/src/shared/configuration.ts
+++ b/packages/karma-typescript/src/shared/configuration.ts
@@ -34,6 +34,7 @@ export class Configuration implements KarmaTypescriptConfig {
     public reports: Reports;
     public transformPath: (filepath: string) => string;
     public tsconfig: string;
+    public stopOnFailure: boolean;
 
     public hasCoverageThreshold: boolean;
 
@@ -186,6 +187,7 @@ export class Configuration implements KarmaTypescriptConfig {
         this.tsconfig = this.karmaTypescriptConfig.tsconfig;
         this.assertExtendable("exclude");
         this.assertExtendable("include");
+        this.stopOnFailure = this.karmaTypescriptConfig.stopOnFailure || true;
     }
 
     private configurePreprocessor() {

--- a/packages/karma-typescript/src/shared/configuration.ts
+++ b/packages/karma-typescript/src/shared/configuration.ts
@@ -187,7 +187,7 @@ export class Configuration implements KarmaTypescriptConfig {
         this.tsconfig = this.karmaTypescriptConfig.tsconfig;
         this.assertExtendable("exclude");
         this.assertExtendable("include");
-        this.stopOnFailure = this.karmaTypescriptConfig.stopOnFailure || true;
+        this.stopOnFailure = this.karmaTypescriptConfig.stopOnFailure !== false;
     }
 
     private configurePreprocessor() {

